// react-markdown-textarea/lib/index.js
// Generated by CoffeeScript undefined
var React, Textarea, marked,
  __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

React = require('react/addons');

if (typeof window !== "undefined" && window !== null) {
  window.React = React;
}

Textarea = require('react-textarea-autosize');

marked = require('marked');

module.exports = React.createClass({
  displayName: 'MarkdownTextarea',
  getDefaultProps: function() {
    return {
      buttonText: "Save",
      onSave: function(value) {},
      onDelete: function() {},
      onChange: function(value) {},
      deleteButton: false,
      spinnerOptions: {}
    };
  },
  getInitialState: function() {
    var state;
    state = {
      active: 'write',
      value: this.props.initialValue != null ? this.props.initialValue : ""
    };
    return state;
  },
  render: function() {
    var cx, liStyles, previewTabClasses, tabs, textarea, textareaStyles, ulStyles, writeTabClasses;
    cx = React.addons.classSet;
    writeTabClasses = cx({
      'react-markdown-textarea__nav__item': true,
      'react-markdown-textarea__nav__item--active': this.state.active === "write"
    });
    previewTabClasses = cx({
      'react-markdown-textarea__nav__item': true,
      'react-markdown-textarea__nav__item--active': this.state.active === "preview"
    });
    ulStyles = {
      display: 'inline-block'
    };
    liStyles = {
      listStyle: 'none',
      float: 'left',
      cursor: 'pointer'
    };
    textareaStyles = {
      display: 'block',
      resize: 'none'
    };
    if (this.state.active === 'write') {
      textarea = this.transferPropsTo(Textarea({
        "className": "react-markdown-textarea__textarea",
        "value": this.state.value,
        "onChange": this.handleChange,
        "ref": "textarea",
        "style": textareaStyles
      }));
    } else {
      textarea = React.DOM.div({
        "className": "react-markdown-textarea__preview",
        "dangerouslySetInnerHTML": {
          __html: marked(this.state.value)
        }
      });
    }
    if (!this.props.noPreview) {
      tabs = React.DOM.ul({
        "className": "react-markdown-textarea__nav",
        "onMouseDown": this.toggleTab,
        "style": ulStyles
      }, React.DOM.li({
        "className": writeTabClasses,
        "style": liStyles
      }, "Write"), React.DOM.li({
        "className": previewTabClasses,
        "style": liStyles
      }, "Preview"));
    }
    return React.DOM.div({
      "className": "react-markdown-textarea"
    }, tabs, React.DOM.div({
      "className": "react-markdown-textarea__textarea-wrapper"
    }, textarea, React.DOM.button({
      "onClick": this._onSave,
      "disabled": (this.props.saving ? "disabled" : false),
      "className": "react-markdown-textarea__save-button"
    }, this.props.buttonText), (this.props.deleteButton ? React.DOM.button({
      "onClick": this._onDelete,
      "disabled": (this.props.saving ? "disabled" : false),
      "className": "react-markdown-textarea__delete-button"
    }, "Delete") : void 0), (this.props.saving && (this.props.spinner != null) ? this.props.spinner(this.props.spinnerOptions) : void 0)));
  },
  toggleTab: function(e) {
    if (e.target.tagName !== "LI") {
      return;
    }
    if (__indexOf.call(e.target.className.split(/\s+/), "react-markdown-textarea__nav__item--active") >= 0) {
      return;
    }
    if (this.state.active === "write") {
      return this.setState({
        active: 'preview'
      });
    } else {
      return this.setState({
        active: 'write'
      });
    }
  },
  handleChange: function(e) {
    var newValue;
    newValue = this.refs.textarea.getDOMNode().value;
    this.setState({
      value: newValue
    });
    return this.props.onChange(newValue);
  },
  _onSave: function() {
    return this.props.onSave(this.state.value);
  },
  _onDelete: function() {
    return this.props.onDelete();
  }
});
